### PR TITLE
jffs2: apply libmtd and libstorage interface

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

 - remove associations with imx6ull
 - mtd interface is moved to generic library
 - remove errno.h
 - remove magic.h and move JFFS2_SUPER_MAGIC to phoenix-rtos.h
 - some types definitions have been moved to libmtd or removed

**NOTE!**
The core implementation of the `libjffs2.c` remained unchanged - it is not a part of this PR.

**TODO in the upcoming PRs:**
 - testing NAND flash memory - (changes in `jffs2/phoenix-rtos/mtd.c`)
 - umount  

DTR-102, PD-216

## Motivation and Context

The MTD interface allows to use the `jffs2` filesystem as a generic library not associated with any board. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zynq7000).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (https://github.com/phoenix-rtos/phoenix-rtos-corelibs/pull/12).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: jffs2/libjffs2.c [hunk 1]
```diff
@@ -363,7 +366,7 @@ static int jffs2_srv_link(jffs2_partition_t *p, oid_t *dir, const char *name, oi
 	if (name == NULL || !strlen(name))
 		return -EINVAL;
```
Comment (kemonats) [inline jffs2/libjffs2.c:367]: Besides PR
```suggestion
	if (name == NULL || (*name == '\0'))
		return -EINVAL;
```

File: jffs2/libjffs2.c [hunk 2]
```diff
@@ -496,20 +506,30 @@ static int jffs2_srv_unlink(jffs2_partition_t *p, oid_t *dir, const char *name)
...
+		return -EINVAL;
+
+	p = (jffs2_partition_t *)info;
+	c = JFFS2_SB_INFO(p->sb);
 
 	if (jffs2_is_readonly(c))
 		return -EROFS;
 
-	if (name == NULL || !strlen(name))
+	if (name == NULL || !namelen)
```
Comment (kemonats) [inline jffs2/libjffs2.c:529]: Here you check `name == NULL` and before that you call `strlen(name)`.

File: jffs2/libjffs2.c [hunk 3]
```diff
@@ -530,7 +550,7 @@ static int jffs2_srv_create(jffs2_partition_t *p, oid_t *dir, const char *name,
 	dentry = malloc(sizeof(struct dentry));
 	memset(dentry, 0, sizeof(struct dentry));
```
Comment (kemonats) [inline jffs2/libjffs2.c:551]: Besides PR.
No checking of the value returned by the `malloc()` function.
Maybe instead  `malloc(), memset()` use `calloc()`.

File: jffs2/libjffs2.c [hunk 4]
```diff
@@ -568,7 +588,7 @@ static int jffs2_srv_create(jffs2_partition_t *p, oid_t *dir, const char *name,
 			break;
 		case otSymlink:
 			/* empty target check */
-			if (dentry->d_name.len >= (namelen - 1) || !strlen(name + dentry->d_name.len + 1)) {
+			if (dentry->d_name.len >= namelen || !(namelen + dentry->d_name.len + 1)) {
```
Comment (kemonats) [inline jffs2/libjffs2.c:591]:
```
size_t namelen;
uint32_t dentry_d_name.len;
and
namelen + dentry->d_name.len + 1 - never == 0
```

File: jffs2/libjffs2.c [hunk 5]
```diff
@@ -58,17 +44,23 @@ struct inode *jffs2_srv_get(jffs2_partition_t *p, oid_t *oid)
...
 	struct dentry *dentry, *dtemp;
 	struct inode *inode = NULL;
 	int len = 0;
 	char *end;
+	jffs2_partition_t *p;
+
+	if (info == NULL)
+		return -EINVAL;
+
+	p = (jffs2_partition_t *)info;
```
Comment (agkaminski) [inline jffs2/libjffs2.c:58]: Why? Can't we just accept `jffs2_partition_t *info` as a parameter?

File: jffs2/libjffs2.c [hunk 6]
```diff
@@ -918,236 +978,85 @@ static int jffs2_srv_statfs(jffs2_partition_t *p, void *buf, size_t len)
...
-int jffs2lib_mount_partition(void *partition)
-{
-	jffs2_partition_t *p = (jffs2_partition_t *)partition;
-
-	object_init(p);
-	dev_init(&p->devs);
-
-	if (jffs2_common.fs->mount(jffs2_common.fs, 0, "jffs2", p) == NULL)
+	if (jffs2_common.fs->mount(jffs2_common.fs, 0, "jffs2", part) == NULL) {
+		printf("jffs2: Failed to mount partition.\n");
```
Comment (agkaminski) [inline jffs2/libjffs2.c:1050]: `part` is leaked here

**Review by agkaminski — COMMENTED** [review]
Some remarks applies to the whole PR

File: jffs2/libjffs2.c (hunk 5 above)
Comment (agkaminski) [reply jffs2/libjffs2.c:58]: Ok, I think I understand, this is API, and we cannot export this type. If that's right please just resolve this remark

File: jffs2/libjffs2.c [hunk 7]
```diff
@@ -58,17 +44,23 @@ struct inode *jffs2_srv_get(jffs2_partition_t *p, oid_t *oid)
...
-static int jffs2_srv_lookup(jffs2_partition_t *p, oid_t *dir, const char *name, oid_t *res, oid_t *dev, char *lnk, int lnksz)
+static int libjffs2_lookup(void *info, oid_t *dir, const char *name, oid_t *res, oid_t *dev, char *lnk, int lnksz)
 {
 	struct dentry *dentry, *dtemp;
 	struct inode *inode = NULL;
 	int len = 0;
 	char *end;
+	jffs2_partition_t *p;
+
+	if (info == NULL)
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply jffs2/libjffs2.c:55]: Ok, I think I understand, this is API, so we need to check params. If that's right please just resolve this remark

File: jffs2/libjffs2.c (hunk 5 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c:58]: It is a generic interface used by the other filesystems as well.

File: jffs2/libjffs2.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (HBuczynski) [reply jffs2/libjffs2.c:55]: Yes, we have to check parameters.

File: jffs2/libjffs2.c [hunk 8]
```diff
@@ -496,20 +506,30 @@ static int jffs2_srv_unlink(jffs2_partition_t *p, oid_t *dir, const char *name)
...
 {
 	struct inode *idir, *inode;
 	struct dentry *dentry, *dtemp;
 	int ret = 0;
-	struct jffs2_sb_info *c = JFFS2_SB_INFO(p->sb);
+	struct jffs2_sb_info *c;
+	jffs2_partition_t *p;
+
+	size_t namelen = strlen(name);
+
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Done

File: jffs2/libjffs2.c [hunk 9]
```diff
@@ -208,7 +207,7 @@ static int jffs2_srv_setattr(jffs2_partition_t *p, oid_t *oid, int type, long lo
 
 		case (atMode): /* mode */
 			iattr.ia_valid = ATTR_MODE;
-			iattr.ia_mode = (inode->i_mode & ~ALLPERMS) | (attr & ALLPERMS);
+			iattr.ia_mode = (inode->i_mode & ~0xffff) | (attr & 0xffff);
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply jffs2/libjffs2.c]: I agree with, I am surprised that this change has been published. I suppose that it is some copy-paste residue, cause I wasn't going to change it.

File: jffs2/libjffs2.c [hunk 10]
```diff
@@ -918,236 +978,85 @@ static int jffs2_srv_statfs(jffs2_partition_t *p, void *buf, size_t len)
...
-			continue;
-		}
-
-		beginthread(jffs2_mount_partition, 3, jffs2_common.partition[i].stack, 0x1000, &jffs2_common.partition[i]);
-	}
-
-	delayed_work_starter(system_long_wq);
-
-	return EOK;
+	return -EINVAL;
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Done

File: jffs2/libjffs2.c (hunk 2 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Thanks for notice. Done.

File: jffs2/libjffs2.c [hunk 11]
```diff
@@ -530,7 +550,7 @@ static int jffs2_srv_create(jffs2_partition_t *p, oid_t *dir, const char *name,
 	dentry = malloc(sizeof(struct dentry));
 	memset(dentry, 0, sizeof(struct dentry));
 	dentry->d_name.name = strdup(name);
-	dentry->d_name.len = strlen(name);
+	dentry->d_name.len = namelen;
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Done

File: jffs2/libjffs2.c (hunk 4 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Done

File: jffs2/libjffs2.c [hunk 12]
```diff
@@ -185,13 +177,20 @@ static int jffs2_srv_lookup(jffs2_partition_t *p, oid_t *dir, const char *name,
...
-static int jffs2_srv_setattr(jffs2_partition_t *p, oid_t *oid, int type, long long attr, void *data, ssize_t size)
+static int libjffs2_setattr(void *info, oid_t *oid, int type, long long attr, void *data, size_t size)
 {
 	struct iattr iattr;
 	struct inode *inode;
 	struct dentry dentry;
 	int ret;
-	struct jffs2_sb_info *c = JFFS2_SB_INFO(p->sb);
+	struct jffs2_sb_info *c;
+	jffs2_partition_t *p;
```
Comment (agkaminski) [inline jffs2/libjffs2.c:187]: You can safely do (it's not a dereference):
```suggestion
	jffs2_partition_t *p = (jffs2_partition_t *)info;
```

**Review by agkaminski — COMMENTED** [review]
Small nitpick applying to the whole PR. Otherwise I see some thinks in need of repair (MISRA violations, not checking malloc retval on beginthread), but these aren't in scope of this PR.

File: jffs2/libjffs2.c (hunk 12 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c:187]: Done

File: jffs2/libjffs2.c [hunk 13]
```diff
@@ -58,17 +44,23 @@ struct inode *jffs2_srv_get(jffs2_partition_t *p, oid_t *oid)
...
 
 
-static int jffs2_srv_lookup(jffs2_partition_t *p, oid_t *dir, const char *name, oid_t *res, oid_t *dev, char *lnk, int lnksz)
+static int libjffs2_lookup(void *info, oid_t *dir, const char *name, oid_t *res, oid_t *dev, char *lnk, int lnksz)
 {
 	struct dentry *dentry, *dtemp;
 	struct inode *inode = NULL;
 	int len = 0;
 	char *end;
+	jffs2_partition_t *p;
```
Comment (agkaminski) [inline jffs2/libjffs2.c:53]: Missed one

File: jffs2/libjffs2.c [hunk 14]
```diff
@@ -918,236 +975,86 @@ static int jffs2_srv_statfs(jffs2_partition_t *p, void *buf, size_t len)
...
-		break;
-
-	case mtCreate:
-		msg->o.create.err = jffs2_srv_create(p, &msg->i.create.dir, msg->i.data, msg->i.size, &msg->o.create.oid, msg->i.create.type, msg->i.create.mode, &msg->i.create.dev);
-		break;
-
-	case mtDestroy:
-		msg->o.io.err = jffs2_srv_destroy(&msg->i.destroy.oid);
-		break;
+	return -EINVAL;
```
Comment (agkaminski) [inline jffs2/libjffs2.c:980]: I missed this one, maybe ENOSYS here too?

File: jffs2/libjffs2.c (hunk 13 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c]: Yeah, fixed.

File: jffs2/libjffs2.c (hunk 14 above)
Comment (HBuczynski) [reply jffs2/libjffs2.c]: ok

**Review by agkaminski — APPROVED** [review]

File: jffs2/libjffs2.c [hunk 15]
```diff
@@ -530,7 +550,10 @@ static int jffs2_srv_create(jffs2_partition_t *p, oid_t *dir, const char *name,
 	dentry = malloc(sizeof(struct dentry));
 	memset(dentry, 0, sizeof(struct dentry));
 	dentry->d_name.name = strdup(name);
-	dentry->d_name.len = strlen(name);
+	if (dentry->d_name.name == NULL)
+		return -ENOMEM;
```
Comment (kemonats) [inline jffs2/libjffs2.c:554]: Memory leak: `dentry`.
```suggestion
	if (dentry->d_name.name == NULL) {
		free(dentry);
		return -ENOMEM;
	}
```

Comment (HBuczynski) [reply jffs2/libjffs2.c]: Thanks, done.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**